### PR TITLE
Add Prompt2Prompt pipeline

### PR DESCRIPTION
**Pull request by UmerHA**
# What does this PR do?
Adds the Prompt-to-Prompt pipeline, as discussed in #2121

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

Pipelines:  @patrickvonplaten and @sayakpaul

### CODE REVIEW

**Comment by UmerHA** [comment]
Hey everyone, 

I have a few questions still. I'll copy them from the issue thread here:

1. Currently, the source and target images are produced simultaneously. In real-life usage, I would imagine they'd be created sequentially. I'd first create a source image, iterate through that until I'm happy with it, and only then edit it. This implies (a) saving the source image's attention maps to the pipe output object, and (b) accepting it as an optional parameter in __call__. Should we do that?

2. Is integration testing with stable-diffusion-v1-4 and size 512 x 512 okay, or would you choose a smaller model / shape?

3. Should we add code to visualize the averaged attention masks? Would that be used in practice? For reference, the paper and @Weifeng-Chen's code include that.

4. (new question) Should we add a demo to the HF hub and link to that in the documentation?

Thanks!

(cc @patrickvonplaten)

**Comment by sayakpaul** [comment]
> (new question) Should we add a demo to the HF hub and link to that in the documentation?

Sure. I think we can also provide a GPU grant. 

> Should we add code to visualize the averaged attention masks? Would that be used in practice? For reference, the paper and @Weifeng-Chen's code include that.

I think we can do that in a GitHub gist and mention it from the docs. Does that work?

> Is integration testing with stable-diffusion-v1-4 and size 512 x 512 okay, or would you choose a smaller model / shape?

For now, I think we can just do fast tests. 

> Currently, the source and target images are produced simultaneously. In real-life usage, I would imagine they'd be created sequentially. I'd first create a source image, iterate through that until I'm happy with it, and only then edit it. This implies (a) saving the source image's attention maps to the pipe output object, and (b) accepting it as an optional parameter in call. Should we do that?

This is a good point, but in the interest of simplicity, I would prefer to go with as is.

**Comment by UmerHA** [comment]
> Sure. I think we can also provide a GPU grant.

That would be great! I have created a dummy space and written the link into the docs. As soon as the PR is merged ( & the classes are available in diffusers), I'll build the space.
 
> I think we can do that in a GitHub gist and mention it from the docs. Does that work?

Works & done.

> For now, I think we can just do fast tests.

I already created the integration tests. I'd leave it at that, unless you'd like me to delete them.

> This is a good point, but in the interest of simplicity, I would prefer to go with as is.

Alright.

**Comment by UmerHA** [comment]
Hi @DN6 @yiyixuxu, pinging you as the PR has been stale for a week. :)

**Comment by DN6** [comment]
Hi @UmerHA. I should be able to review this by tomorrow.

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py
Comment (DN6) [inline src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py:1]: Is it possible to remove this file from the PR. It isn't relevant.

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion_img2img.py
Comment (DN6) [inline src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion_img2img.py:1]: Is it possible to remove this file from the PR. It isn't relevant.

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py
Comment (UmerHA) [reply src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py:1]: Have removed the automatic formatting on all 3 unrelated files

**Comment by patrickvonplaten** [comment]
I'm not sure how relevant this pipeline still is to be honest - @apolinario do you think there is still a big demand for this pipeline?

**Comment by Zeqiang-Lai** [comment]
As an user of this pipeline, I hope this pipeline could be merged or at least make it a community pipeline.

**Comment by patrickvonplaten** [comment]
@UmerHA,

Do you think we could add this pipeline as a community pipeline instead? Our test suite and maintenance cost is exploding at the moment and we need to be a bit more selective which pipelines shall be added to "main"

**Comment by UmerHA** [comment]
@patrickvonplaten Absolutely, will do!

**Comment by UmerHA** [comment]
> @UmerHA,
> 
> Do you think we could add this pipeline as a community pipeline instead? Our test suite and maintenance cost is exploding at the moment and we need to be a bit more selective which pipelines shall be added to "main"

Have now:
- moved the pipeline to community
- deleted the tests
- condensed the docs and moved them to community

I think I've done everything, but would appreciate if you could quickly check. Thanks!

File: src/diffusers/utils/dummy_torch_and_transformers_objects.py [hunk 1]
```diff
@@ -512,6 +512,21 @@ def from_pretrained(cls, *args, **kwargs):
...
+
+    @classmethod
+    def from_config(cls, *args, **kwargs):
+        requires_backends(cls, ["torch", "transformers"])
+
+    @classmethod
+    def from_pretrained(cls, *args, **kwargs):
+        requires_backends(cls, ["torch", "transformers"])
+
+
```
Comment (patrickvonplaten) [inline src/diffusers/utils/dummy_torch_and_transformers_objects.py:529]:
```suggestion
```

**Review by patrickvonplaten — COMMENTED** [review]
Just some final clean-ups :-)

**Comment by keturn** [comment]
As this has been an InvokeAI feature since before Invoke adopted diffusers, it's been on the list of things I've hoped to see supported here.

Partly to ease Invoke's maintenance for a feature that has required some messy kludges at times (factoring out the AttentionProcessor helped, but I lost track of where that whole thing ended up), but also because Invoke's implementation only does the **replace** operator and I've been looking forward to being able to use **reweight**.

> * deleted the tests

😵 a fully implemented feature with tests and then you _deleted_ them? _[keturn dies inside]_

**Comment by patrickvonplaten** [comment]
> > * deleted the tests

Tests are exploding and becoming too expensive/slow (currently at 20min just for fast tests). We sadly need to be more selective with pipeline testing.

**Comment by patrickvonplaten** [comment]
Thanks a mille for the PR @UmerHA

**Comment by keturn** [comment]
I've opened an issue over at https://github.com/invoke-ai/InvokeAI/issues/4541 to figure out how to handle things on Invoke's side. @UmerHA, if you're interested in contributing prompt2prompt and attention-map-visualization code to Invoke, your input would be very welcome!

**Comment by betterze** [comment]
How can I use this? I try 

```
from diffusers.pipelines import Prompt2PromptPipeline
```

it outputs 

```
ImportError: cannot import name 'Prompt2PromptPipeline' from 'diffusers.pipelines' (/opt/venv/lib/python3.10/site-packages/diffusers/pipelines/__init__.py)
```

I have the newest version of diffusers.

**Comment by patrickvonplaten** [comment]
@betterze,

can you try:

```py
from diffusers import DiffusionPipeline

pipeline = DiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", custom_pipeline="pipeline_prompt2prompt")
```

**Comment by betterze** [comment]
Does it support sdxl? when i run 

```

import torch 
from pipeline_prompt2prompt import Prompt2PromptPipeline
pipe = Prompt2PromptPipeline.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0"   # stabilityai/sd-turbo, runwayml/stable-diffusion-v1-5
)
pipe = pipe.to("cuda")

prompts = ['a glass of cocktail on the table',
           'a glass of white on the table']


# prompts = ['a sad face',
           # 'a happy face']

cross_attention_kwargs = {
    "edit_type": "replace",
    "n_cross_replace": 1.0,
    # "n_self_replace": 1.0,
    "local_blend_words": ["cocktail", "white"]
    }

generator = torch.Generator(device="cuda").manual_seed(1)
outputs = pipe(prompt=prompts,generator=generator, height=512, width=512, num_inference_steps=50,guidance_scale=7.0, cross_attention_kwargs=cross_attention_kwargs)

```

the following error occurs

```
Cell In[3], line 16
      8 cross_attention_kwargs = {
      9     "edit_type": "replace",
     10     "n_cross_replace": 1.0,
     11     # "n_self_replace": 1.0,
     12     "local_blend_words": ["cocktail", "white"]
     13     }
     15 generator = torch.Generator(device="cuda").manual_seed(1)
---> 16 outputs = pipe(prompt=prompts,generator=generator, height=512, width=512, num_inference_steps=50,guidance_scale=7.0, cross_attention_kwargs=cross_attention_kwargs)
     17 # outputs = pipe(prompt=prompts,generator=generator, height=512, width=512, num_inference_steps=1,guidance_scale=0.0, cross_attention_kwargs=cross_attention_kwargs,latents=inv_latents)
     19 from matplotlib import pyplot as plt

File /opt/venv/lib/python3.10/site-packages/torch/utils/_contextlib.py:115, in context_decorator.<locals>.decorate_context(*args, **kwargs)
    112 @functools.wraps(func)
    113 def decorate_context(*args, **kwargs):
    114     with ctx_factory():
--> 115         return func(*args, **kwargs)

File /sensei-fs/tenants/Sensei-AdobeResearchTeam/share-zongzew/clio3/pipeline_prompt2prompt.py:245, in Prompt2PromptPipeline.__call__(self, prompt, height, width, num_inference_steps, guidance_scale, negative_prompt, num_images_per_prompt, eta, generator, latents, prompt_embeds, negative_prompt_embeds, output_type, return_dict, callback, callback_steps, cross_attention_kwargs, guidance_rescale)
    242 latent_model_input = self.scheduler.scale_model_input(latent_model_input, t)
    244 # predict the noise residual
--> 245 noise_pred = self.unet(latent_model_input, t, encoder_hidden_states=prompt_embeds).sample
    247 # perform guidance
    248 if do_classifier_free_guidance:

File /opt/venv/lib/python3.10/site-packages/torch/nn/modules/module.py:1501, in Module._call_impl(self, *args, **kwargs)
   1496 # If we don't have any hooks, we want to skip the rest of the logic in
   1497 # this function, and just call forward.
   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1499         or _global_backward_pre_hooks or _global_backward_hooks
   1500         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1501     return forward_call(*args, **kwargs)
   1502 # Do not call functions when jit is used
   1503 full_backward_hooks, non_full_backward_hooks = [], []

File /opt/venv/lib/python3.10/site-packages/diffusers/models/unet_2d_condition.py:967, in UNet2DConditionModel.forward(self, sample, timestep, encoder_hidden_states, class_labels, timestep_cond, attention_mask, cross_attention_kwargs, added_cond_kwargs, down_block_additional_residuals, mid_block_additional_residual, down_intrablock_additional_residuals, encoder_attention_mask, return_dict)
    964     aug_emb = self.add_embedding(text_embs, image_embs)
    965 elif self.config.addition_embed_type == "text_time":
    966     # SDXL - style
--> 967     if "text_embeds" not in added_cond_kwargs:
    968         raise ValueError(
    969             f"{self.__class__} has the config param `addition_embed_type` set to 'text_time' which requires the keyword argument `text_embeds` to be passed in `added_cond_kwargs`"
    970         )
    971     text_embeds = added_cond_kwargs.get("text_embeds")

TypeError: argument of type 'NoneType' is not iterable
```

Thank you for your help.

**Comment by UmerHA** [comment]
@betterze unfortunately no. SDXL versions of pipelines are their own classes. For this pipeline, there currently is no SDXL version.

Feel free to implement one!

**Pull request merged**